### ISSUE TITLE
too-many-function-args differs with no-member for multiple return types

### ISSUE THREAD

**Issue by AlexRiina** [PyCQA/pylint#4170]
This is a bit of a twist on some other bug reports about conditional response types. In this report, it seems like `no-member` and `too-many-function-args` disagree on when to raise a error even though it seems like either both should or both shouldn't. The example is slightly more elaborate that a minimal example to make it clearer in what case this might show up.

### Steps to reproduce

Given a file like the below where 

```
class X:
    SHARED = 'x'
    X_SOLO = 'x'

    def hi(self, arg1):
        pass


class Y:
    SHARED = 'y'
    Y_SOLO = 'y'

    def hi(self, arg1, arg2):
        pass


def either(a):
    if a == 'X':
        return X
    elif a == 'Y':
        return Y


def called_with_x(a):
    x = either(a)()
    x.SHARED
    x.X_SOLO
    x.hi('hey')


def called_with_y(a):
    y = either(a)()
    y.SHARED
    y.Y_SOLO
    y.hi('hey', 'there')
```

### Current behavior
Pylint currently finds
```
E1121: Too many positional arguments for method call (too-many-function-args)
```
on the line with `hi('hey', 'there')`

### Expected behavior
From my understanding, pylint cannot tell which type will be returned by `either(a)` so the inference engine attempts to validate all of the types. With `no-member`, lines like `x.X_SOLO` and `y.Y_SOLO` should raise errors because one type doesn't have such a value but it appears to pass if any possible type can be validated. On the other hand, the `too-many-function-args` error does trigger which means it raises it's warning if any type fails validation.

I'm working on a python 2 codebase and noticed that in pylint==1.9.5 astroid==1.6.6, `too-many-function-args` error is not raised but in recent pylint and astroid the error is raised.

### pylint --version output

Result of `pylint --version` output:
```
pylint --version
pylint 2.7.2
astroid 2.5.1
Python 3.8.5 (default, Oct 28 2020, 17:06:51) 
[GCC 7.5.0]
```

**Comment by Pierre-Sassoulas** [comment]
Thank you for the very good example and bug report. I think there is two clashing philosophy here, those who want pylint to raise a message for anything suspicious and triage it themself after the fact, and those who don't want to have too much false positives because it's making pylint untrustworthy. In any case there is an inconsistency, I think both school will agree that those two lines should raise or not raise in your example.

**Comment by DanielNoord** [comment]
I don't think we will ever solve this. This is dependent on control flow inference, but even if we had full flow inference this would be tricky.
Both `pyright` and `mypy` don't understand this currently as the flow is so inexplicit. I would suggest to use an `isinstance` check instead of the `Y_SOLO` check. We don't support those yet, but there are multiple issues tracking our progress towards that 😄
